File: frontend/survey-app/src/app/services/survey.service.ts
// Service class for handling HTTP requests to the survey REST API
// Done by Shreyas Patil
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Survey } from '../models/survey.model';

@Injectable({
  providedIn: 'root',
})
export class SurveyService {
  private apiUrl = 'http://localhost:8080/api/surveys';

  constructor(private http: HttpClient) {}

  /**
   * Get all surveys
   */
  getAllSurveys(): Observable<Survey[]> {
    return this.http.get<Survey[]>(this.apiUrl);
  }

  /**
   * Get a specific survey by ID
   */
  getSurveyById(id: number): Observable<Survey> {
    return this.http.get<Survey>(`${this.apiUrl}/${id}`);
  }

  /**
   * Create a new survey
   */
  createSurvey(survey: Survey): Observable<Survey> {
    return this.http.post<Survey>(this.apiUrl, survey);
  }

  /**
   * Update an existing survey
   */
  updateSurvey(id: number, survey: Survey): Observable<Survey> {
    return this.http.put<Survey>(`${this.apiUrl}/${id}`, survey);
  }

  /**
   * Delete a survey
   */
  deleteSurvey(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/survey-app/src/main/java/edu/gmu/swe642/surveyapp/service/SurveyService.java
package edu.gmu.swe642.surveyapp.service;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import edu.gmu.swe642.surveyapp.model.Survey;
import edu.gmu.swe642.surveyapp.repository.SurveyRepository;

/**
 * Done by Sri Bhuvan, Praneeth Naidu, Ankit Raut, Shreyas Patil
 * Service class for Survey entity operations
 */
@Service
public class SurveyService {

    private final SurveyRepository surveyRepository;

    @Autowired
    public SurveyService(SurveyRepository surveyRepository) {
        this.surveyRepository = surveyRepository;
    }

    /**
     * Retrieve all surveys
     */
    public List<Survey> getAllSurveys() {
        return surveyRepository.findAll();
    }

    /**
     * Retrieve a specific survey by ID
     */
    public Optional<Survey> getSurveyById(Long id) {
        return surveyRepository.findById(id);
    }

    /**
     * Create a new survey
     */
    public Survey createSurvey(Survey survey) {
        return surveyRepository.save(survey);
    }

    /**
     * Update an existing survey
     */
    public Survey updateSurvey(Long id, Survey surveyDetails) {
        Survey survey = surveyRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Survey not found with id: " + id));
        
        survey.setFirstName(surveyDetails.getFirstName());
        survey.setLastName(surveyDetails.getLastName());
        survey.setStreetAddress(surveyDetails.getStreetAddress());
        survey.setCity(surveyDetails.getCity());
        survey.setState(surveyDetails.getState());
        survey.setZip(surveyDetails.getZip());
        survey.setPhone(surveyDetails.getPhone());
        survey.setEmail(surveyDetails.getEmail());
        survey.setSurveyDate(surveyDetails.getSurveyDate());
        survey.setLikes(surveyDetails.getLikes());
        survey.setInterestSource(surveyDetails.getInterestSource());
        survey.setRecommendationLikelihood(surveyDetails.getRecommendationLikelihood());
        survey.setComments(surveyDetails.getComments());
        
        return surveyRepository.save(survey);
    }

    /**
     * Delete a survey
     */
    public void deleteSurvey(Long id) {
        Survey survey = surveyRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Survey not found with id: " + id));
        surveyRepository.delete(survey);
    }
}